source/_posts/2016-01-05-improvements-to-error-responses-in-rails-5-api-mode.markdown

```diff
@@ -0,0 +1,160 @@
+
+<!--more-->
+
+Before digging into how error responses were improved, let's recap how response format is determined in Rails.
+
+## Response format calculation in Rails
+
+In Rails, the concept of _format_ appears just from the first scaffold command. For example, if you run a simple `bin/rails g scaffold post title`, you can find references to response formats in the controller and also in the routes.
+
+Part of the generated controller looks like the following:
+
+```ruby
+class PostsController < ApplicationController
+
+  ...
+
+  # GET /posts
+  # GET /posts.json
+  def index
+    @posts = Post.all
+  end
+
+  ...
+
+  # POST /posts
+  # POST /posts.json
+  def create
+    @post = Post.new(post_params)
+
+    respond_to do |format|
+      if @post.save
+        format.html { redirect_to @post, notice: 'Post was successfully created.' }
+        format.json { render :show, status: :created, location: @post }
+      else
+        format.html { render :new }
+        format.json { render json: @post.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+end
+``` 
+
+In Rails controllers, you have the ability to write code that can respond in different formats. By default, generated code from the scaffold command can respond on HTML or JSON formats. That is easy to see in the `create` action, but it also happens in the `index` action where the render is implicit (this action returns a list of posts and it's rendered in the appropriate format).
+
+Of course, there should be some sort of rule that determines which format is picked to respond the request. In order to understand those rules, let's now take a look at the routes generated by the scaffold.
+
+The following is the output of `bin/rake routes` in our example application:
+
+```
+   Prefix Verb   URI Pattern               Controller#Action
+    posts GET    /posts(.:format)          posts#index
+          POST   /posts(.:format)          posts#create
+ new_post GET    /posts/new(.:format)      posts#new
+edit_post GET    /posts/:id/edit(.:format) posts#edit
+     post GET    /posts/:id(.:format)      posts#show
+          PATCH  /posts/:id(.:format)      posts#update
+          PUT    /posts/:id(.:format)      posts#update
+          DELETE /posts/:id(.:format)      posts#destroy
+```
+
+These routes are what you get when you have something like `resources: :post` in your routes file. We can see that _format_ here is a dynamic segment present in every generated routes, and this is the most common way in Rails to decide which is the format used to form a response. So, in many cases the format depends on the extension in the request URL, usually appended by a client application (or client side code if it's a Rails default stack application) to indicate the expected format.
```

Comment:
the usually appended part is what is weird, client side could append but I could also request posts.html by writing that in the browser url. So unsure exactly what you're trying to say there